pkg/heartbeat/types/types.go

```diff
@@ -67,3 +72,53 @@ func (d Distribution) String() string {
 	}
 	return "unknown"
 }
+
+type InstanceReport struct {
+	Events []InstanceReportEvent `json:"events"`
+}
+
+type InstanceReportEvent struct {
+	ReportedAt                int64  `json:"reported_at"`
+	LicenseID                 string `json:"license_id"`
+	InstanceID                string `json:"instance_id"`
+	ClusterID                 string `json:"cluster_id"`
+	AppStatus                 string `json:"app_status,omitempty"`
+	ResourceStates            string `json:"resource_states,omitempty"`
+	K8sVersion                string `json:"k8s_version"`
+	K8sDistribution           string `json:"k8s_distribution,omitempty"`
+	DownstreamChannelID       string `json:"downstream_channel_id,omitempty"`
+	DownstreamChannelSequence int64  `json:"downstream_channel_sequence"`
+	DownstreamChannelName     string `json:"downstream_channel_name,omitempty"`
+}
+
+func (r *InstanceReport) Encode() ([]byte, error) {
+	data, err := json.Marshal(r)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal instance report")
+	}
+	compressedData, err := util.GzipData(data)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to gzip instance report")
+	}
+	encodedData := base64.StdEncoding.EncodeToString(compressedData)
+
+	return []byte(encodedData), nil
+}
+
+func DecodeInstanceReport(encodedData []byte) (*InstanceReport, error) {
```

Comment:
this doesn't feel like it belongs in a `types` package since it doesn't belong to a specific type. my suggestion would be to move this (along with the memory store secret logic) to the `instance_report.go` file and maybe rename `(r *InstanceReport) Encode()` to `EncodeInstanceReport`.